### DISCUSSION TITLE
Scripting language as config file

### DISCUSSION THREAD

**Discussion by dvermd** [polybar/polybar#2945] [Ideas]
Hi, coming back to re-activate #2515, I looked into Lua and it's quite easy to interface with.

The main problem I see with Lua is the lack of list/vector type. Lua supports ints, floats and maps. Lists are encoded as maps with integer keys which may not be continuous.

A Lua script can import other Lua scripts and access env variables.

For this to land, I think:
* some Config API must be adapted
* a migration tool must be provided for the actual configuration file
* maybe, the current parser should remain active for a while

The easiest API between Polybar and Lua is a global variable in Lua queried by C++.

Any thoughts on how to improve the Config API (moving the `get_with_prefix` to a map, define an operator[], ???) ?
Should I start with the Config API changes before adding Lua ?

**Comment by patrick96** [comment]
Thanks for putting time into researching this.

We should think about how we want polybar to be configurable. Introducing lua is a big step and we should be clear about where we want to go with it.

I have been thinking about whether simply supporting a new configuration format may already solve our problems.
Languages like [TOML](https://toml.io/en/), [YAMl](https://yaml.org/), and [jsonnet](https://jsonnet.org/) are very powerful and would remove a lot of the burden we currently have with having to extend the parser for more and more features.

However, I have also been reading [this blog post](https://matt-rickard.com/advanced-configuration-languages-are-wrong) and it basically claims that making the configuration language more and more advanced is wrong and a general purpose programming language should be used instead.
I would tend to agree with this. Even if we were to switch to TOML or similar now, we would probably reach its limits at some point and would be in the same position again.

That being said, I don't know if forcing everyone to completely switch to lua is the best idea.
Most of our users likely do not have much or any programming experience. For them, switching over would be a huge burden.
This would implicitly make polybar only suitable for more advanced users.
Even for advanced users (especially for those with large complicated configurations), switching over would still be a big effort.

Maybe a mix could be possible where configuration on the lua side can be imported into existing config files.

For example:

```dosini
[module/i3]
lua-import = myi3module

[module/text]
type = custom/text
label = ${lua-import:mytextlabel}
```

And on the lua side, we would somehow register a module configuration under the name "myi3module" and a label configuration under the name "mytextlabel".

This way, switching over can be done 1 by 1 for those who want to and others could keep their INI config file, only using lua for more complicated modules.

---

Now for how I imagine configuration on the lua side to look like, at least for any of the current functionality.

If we already do the work to support lua, I don't necessarily want to simply provide the same key-value configuration as in the config file, but more of a structured approach.
For example for a label:

```dosini
label = "This is a label"
label-foreground = #ff0000
label-underline = #00ff00
```

I imagine one of the following:

```lua
lbl = Label:new("This is a label")
lbl:set_foreground("#ff0000")
lbl:set_underline("#00ff00")

module:register_label(lbl)
```

Or maybe

```lua
module:register_label(Label:new{
  text = "This is a label",
  foreground = "#ff0000",
  underline = "#00ff00",
})
```

The main idea is that components like labels (or any other [content types](https://github.com/polybar/polybar/wiki/Formatting#content-types)), modules, or formats (basically anything that currently requires multiple configuration values) are represented as objects and build a hierarchy.
So modules are registered in the bar, labels and formats are registered in modules, etc.

What do you think about this?

How did you envision the lua part to look like?

---

I think we should first try to get the simplest possible interaction to work. Maybe importing a single number from lua into our INI configuration file.
Even if we don't want to support that particular functionality in the future, it should give us some insights into how these kind of integrations will work.

After that we can think about what kind of APIs from lua to polybar and from the config file to lua we want to support.

This is going to be a lot of work and I don't expect you to do all of it alone. But I do want to hear what your vision is for this. 

If we support mixing the config file with lua, it also makes it easier to add lua functionality step by step instead of all at once.

**Comment by dvermd** [comment]
I think you're right on using an existing configuration language may lead to the same problems in the future.

Honestly, I find the current configuration quite complicated/not easy to use and not that explicit.

I'm not a big fan of adding a mix between current config and lua because it will add more complexity and might not easily fix the configuration limitation in a straight way. But I understand your concerns about not disturbing the users too much. And both configuration language could co-exist for a while/forever. The ini one might be limited and not allow the fine grained configuration of the lua one.

For example,this took me quite a while to understand: the name of a module used in a `module-left` is taken after their `[module/xyz]`. This can be misleading because in the documentation modules are named by their type `[module/alsa]` for type `internal/alsa` with an entry `alsa` in the wiki but we also have `[module filesystem]` for type `internal/fs` with an entry `filesystem` in the wiki

Having learn the configuration syntax is already a on the users, making them move to just another (maybe easier) syntax might not be as hard as you think. No one knows and a good documentation will surely help a lot.

Lua has only a [few types](https://www.lua.org/pil/2.html) and I think most of the configuration will be done through cascading tables.

Here is a short example of what a lua config file could look like.

```
i3 = {
    type = "internal/i3",
    -- this syntax is needed for key containing '-' or reserved word
    [ "pin-workspaces" ] = true,
    [ "ws-icon" ] = {
        "1;♚",
        "2;♛",
        "3;♜",
        "4;♝",
        "5;♞",
        default = "♟"
    }
}

bars = {
    mybar = {
        monitor = {
            value = "",
            fallback = "",
            strict = false
        },
        modules = {
            left = {
                i3,
                {
                    type = "internal/fs", 
                    interval = 25, 
                    mount = {"/", "/home"}
                }
            }
        }
    }
}
```

I'll push some example code to read it soon.

**Reply by patrick96** [reply]
Hmm, so more of a json-style format of the current config syntax with better syntax for nesting.

I guess a syntax like that isn't too difficult to learn. And you get a lot of features for free because it's a programming language4. I like it!

In the future this could also be extended by allowing callback functions for certain values to do stuff like custom formatters. This would really push polybar's capabilities to the next level.

**Reply by dvermd** [reply]
Polybar or a user might add some helper functions like in your first comment to create part of the maps more easily but in the end, the config starts at `bars`.

Calling from C++ to lua to run some user defined hooks is a great idea !

For now, the new nesting of modules' definition in the bar leads to an incompatibility in the config API. Once in a module, the configuration is read with the section being the name of the module that is defined as a top level item in the configuration file.
Allowing to nest the module definition directly in the bar, leads to a `config` API change where the modules pass the bar name to the `m_conf.xyz` function.
Are you OK with this API change ?

**Reply by dvermd** [reply]
Or, the `config::make()` function could take a bar name and then there will only be one configuration singleton that all modules would use.

**Reply by dvermd** [reply]
OK, got in into my head. Everything is already in place.

I had some difficulties understanding how factory_util::singleton returned the same class (thus static instance) for both calls `config::make("path/to/config", "bar_name")` and `config::make()` later. 
GDB helped me find this is working as intended and I had to understand more about C++ templates.

TIL more about C++ templates

**Reply by patrick96** [reply]
TBH, I'm not the biggest fan of our singleton patterns. At the very least it should be `config::make(...)` and `config::get()`.

**Reply by dvermd** [reply]
The more experience I gain at development the more I tend to stay away from singletons. 

I think it's possible to get rid of the config singleton by passing the `config` object returned by `config_parser::parse` to `bar::make`, `controller::make` and others.
Would you accept a PR for that ?

Also, things that could be included in the PR
* `config::ignore_key` and `config::warn_deprecated` can be merged with a `replacement` argument defaulted to `string()`
* `deprecated_list` does not seem to be used and could be deleted.

**Reply by patrick96** [reply]
> Would you accept a PR for that?

Yes definitely! Also the other refactorings. The config class is kind of a mess

**Reply by dvermd** [reply]
A similar refactoring may be possible to remove the log singleton. Would you also consider integrating it ?

Also, I think the `unsupported.hpp` file can be removed if we add some `#if` in factory.cpp instead to throw the exception from a custom factory function instead of creating a fake type with a macro and throwing when it is instanciated.
Something like 
```
#if ENABLE_I3
      map_entry<i3_module>(),
#else
      {"i3_module", [](const bar_settings&, string&&, const config&)-> module_t{
        throw application_error("No built-in support for 'internal/i3'");
        return module_t();
        }
      },
#endif
```

**Reply by patrick96** [reply]
I don't mind the logger singleton that much. Having to pass it around all the time will clutter up the code quite a bit. I'm sure the logger would benefit from some refactoring, but I don't think always passing it around is the way to go.

Removing `unsupported.hpp` would certainly be a positive thing though. It's a pain having to always update the unsupported module definition when changing the main module interface.

**Reply by dvermd** [reply]
> Removing `unsupported.hpp` would certainly be a positive thing though. It's a pain having to always update the unsupported module definition when changing the main module interface.

I see many ways to get rid of `unsupported.hpp`:
* In `factory.cpp`, create a second map for unsupported modules, add `#if ENABLE_XXX` in the `factories` map and first test `factories` map then the unsupported one then issue the unknown message
* In `factory.cpp`, add a lambda which throws instead of calling the `get_factory` method as proposed in my previous message
* Move the responsibility of being enable or not to the modules themselves:
  * The modules would have the `#if ENABLE_XXX`,
  * they have access the their `TYPE` member for throwing the right message
  * the `factory.cpp` would stay untouched
  * as the real `xxx_module` is instanciated, the `#if ENABLE_XXX` from `all.hpp`, `bar.cpp` and `controller.cpp` could be removed

Any solution you prefer ?

**Reply by patrick96** [reply]
I think the second one is the only viable solution (the one you previously proposed).

Unsupported modules will not have the right header files and libraries available and won't even compile. That's why their files are only included for compilation (in `src/CMakeLists.txt`) if they are enabled.

**Reply by dvermd** [reply]
I have a working version of the second alternative BUT as the unsupported version of the alsa module is not instanciated, the [alsa_module::TYPE](https://github.com/polybar/polybar/blob/master/src/modules/meta/factory.cpp#L73) is not defined and I had to inline the value.

I have [another prototype](https://github.com/dvermd/polybar/tree/remove_unsupported_hpp_always_module) where: 
* the alsa.cpp is always compiled
* a meta/unsupported_module.hpp is defined
* alsa.hpp and alsa.cpp use ENABLE_ALSA to either compile the current version or an unsupported version that only define TYPE

Which one fits polybar best ?

**Reply by patrick96** [reply]
right now `unsupported.hpp` duplicates the module type string, so It think that's fine. Alternatively we can centrally define all module type strings as static constexpr and use those in `alsa_module::TYPE` and in `factory.cpp`

Your second approach looks like it will produce a lot of extra code and you will have a lot of `#if`s around because you often also need to exclude certain headers.

The first approach seems far superior since it only requires a single `#if ENABLE_ALSA` in `factory.cpp` and can be made very concise

**Comment by dvermd** [comment]
In the ini config, the modules are defined outside of the bars and referenced by name in the bars + alignment.

My lua config example defines the modules directly inside the `bars[barname]["modules"][alignment]` table (`alignment` being "left", "center" or "right".

The configuration API is not able to handle that because bars parameters and modules parameters are accessed by the same functions with section as primary key.
This is due to modules being first level entry in the ini configuration file, but not anymore in the lua configuration file.

The configuration API (and modules code) would need these changes:
* create specific `bar_` functions in `config` class to access bar parameters
* update modules functions in `config` to take an `align` parameter
* add `alignment` to modules so they can use it with the updated functions.

Do you want to make the move to defining modules in the bars and make the relevant changes in the code ? 
Or do you prefer to keep the modules configuration outside of the bar configuration in the lua configuration ?

**Reply by patrick96** [reply]
Regardless of how the lua configuration table ends up looking like, we should start out with replicating the INI structure in lua 1-to-1. So every module is its own top-level entry in the config table and `modules-left`, etc. are just a space separated string.

This way we can, for now, reuse all the config code and just get the lua interactions working.
Once we have that, creating good config APIs should be much faster and hopefully easier.

In the end I do think the module configs should live in a nested list in the bar table, just like you showed.

As for the config API, I imagine it would something like this:

* Config settings are no longer directly looked up by "section" and "key", only by key.
* For modules you would be provided config object with all settings for that module. For INI configs, you could just look up the module section and for lua configs, you would provide the module setting as they were given in the module list.
* This approach is generic enough to work for other settings like labels, progressbars, monitors, etc.
